source/agent/addons/quic/QuicTransportStream.cc

```diff
@@ -0,0 +1,225 @@
+
+using v8::Function;
+using v8::FunctionTemplate;
+using v8::Local;
+using v8::Object;
+using v8::ObjectTemplate;
+using v8::Value;
+
+DEFINE_LOGGER(QuicTransportStream, "QuicTransportStream");
+
+Nan::Persistent<v8::Function> QuicTransportStream::s_constructor;
+
+const int uuidSizeInByte = 16;
+
+QuicTransportStream::QuicTransportStream()
+    : QuicTransportStream(nullptr)
+{
+}
+QuicTransportStream::QuicTransportStream(owt::quic::QuicTransportStreamInterface* stream)
+    : m_stream(stream)
+    , m_contentSessionId()
+    , m_receivedContentSessionId(false)
+    , m_isPiped(false)
+{
+}
+
+QuicTransportStream::~QuicTransportStream()
+{
+    if (!uv_is_closing(reinterpret_cast<uv_handle_t*>(&m_asyncOnContentSessionId))) {
+        uv_close(reinterpret_cast<uv_handle_t*>(&m_asyncOnContentSessionId), NULL);
+    }
+    if (!uv_is_closing(reinterpret_cast<uv_handle_t*>(&m_asyncOnData))) {
+        uv_close(reinterpret_cast<uv_handle_t*>(&m_asyncOnData), NULL);
+    }
+    m_stream->SetVisitor(nullptr);
+}
+
+void QuicTransportStream::OnCanRead()
+{
+    if (!m_receivedContentSessionId) {
+        MaybeReadContentSessionId();
+    } else {
+        SignalOnData();
+    }
+}
+
+void QuicTransportStream::OnCanWrite()
+{
+    ELOG_DEBUG("On can write.");
+}
+
+void QuicTransportStream::OnFinRead()
+{
+    ELOG_DEBUG("On FIN read.");
+}
+
+NAN_MODULE_INIT(QuicTransportStream::init)
+{
+    Local<FunctionTemplate> tpl = Nan::New<FunctionTemplate>(newInstance);
+    tpl->SetClassName(Nan::New("QuicTransportStream").ToLocalChecked());
+    Local<ObjectTemplate> instanceTpl = tpl->InstanceTemplate();
+    instanceTpl->SetInternalFieldCount(1);
+
+    Nan::SetPrototypeMethod(tpl, "write", write);
+    Nan::SetPrototypeMethod(tpl, "addDestination", addDestination);
+    //Nan::SetPrototypeMethod(tpl, "write", write);
+
+    s_constructor.Reset(Nan::GetFunction(tpl).ToLocalChecked());
+    Nan::Set(target, Nan::New("QuicTransportStream").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
+}
+
+NAN_METHOD(QuicTransportStream::newInstance)
+{
+    if (!info.IsConstructCall()) {
+        ELOG_DEBUG("Not construct call.");
+        return;
+    }
+    QuicTransportStream* obj = new QuicTransportStream();
+    obj->Wrap(info.This());
+    uv_async_init(uv_default_loop(), &obj->m_asyncOnContentSessionId, &QuicTransportStream::onContentSessionId);
+    uv_async_init(uv_default_loop(), &obj->m_asyncOnData, &QuicTransportStream::onData);
+    info.GetReturnValue().Set(info.This());
+}
+
+NAN_METHOD(QuicTransportStream::write){
+    if(info.Length()<2){
+        Nan::ThrowTypeError("Data is not provided.");
+        return;
+    }
+    QuicTransportStream* obj = Nan::ObjectWrap::Unwrap<QuicTransportStream>(info.Holder());
+    uint8_t* buffer=(uint8_t*)node::Buffer::Data(info[0]->ToObject());
+    unsigned int length = info[1]->Uint32Value();
+    obj->m_stream->Write(buffer, length);
+    info.GetReturnValue().Set(info.This());
```

Comment:
Understand the "Write" api of QuicTransportStream does not have return value. but How can you let js know if the write succeed or not if you're not returning bytes written?

Reply:
A return value was added to QUIC SDK.